### PR TITLE
First draft of the tutorial.

### PR DESCRIPTION
**Pull request by Natim**
When writting this tutorial I've got issues with the current kinto implementation

 - [x] Groups doesn't looks to be plugged with the user principal yet #78 
 - [x] Default user's buckets #71

You may still start to review the tutorial on both the content and the form of it as well as bits that may be missing.

Please wait for rearranging it (changing titles etc)

### CODE REVIEW

File: docs/tutorial.rst [hunk 1]
```diff
@@ -0,0 +1,981 @@
...
+    $ http GET https://kinto.dev.mozaws.net/v1/buckets/todo/collections/tasks/records/2fa91620-f4fa-412e-aee0-957a7ad2dc0e \
+        -v --auth 'mary:marypassword'
+
+
+
+
+Conclusion
+==========
+
+In this tutoriel, you have see all the concept exposed by Kinto:
```
Comment (ametaireau) [inline docs/tutorial.rst]: tutorial

File: docs/tutorial.rst [hunk 2]
```diff
@@ -0,0 +1,987 @@
+.. _tutorial:
+
+A first app with Kinto
+######################
+
+There are actually two kinds of app that you may want to use Kinto with:
```
Comment (leplatrem) [inline docs/tutorial.rst]: ``of applications where *Kinto* is particularly relevant``

File: docs/tutorial.rst [hunk 3]
```diff
@@ -0,0 +1,987 @@
...
+
+  - App to sync user data between her devices;
```
Comment (leplatrem) [inline docs/tutorial.rst]: - Sync user...

File: docs/tutorial.rst [hunk 4]
```diff
@@ -0,0 +1,987 @@
...
+  - App to sync and share data between users, with fined-grained permissions.
+
+Most of the app will fit in one or the other.
```
Comment (leplatrem) [inline docs/tutorial.rst]: You can get rid of this IMO

File: docs/tutorial.rst [hunk 5]
```diff
@@ -0,0 +1,987 @@
...
+
+
+Sync user data between devices
+==============================
+
+Let say that we want to do a TodoMVC backend that will sync user tasks
```
Comment (leplatrem) [inline docs/tutorial.rst]: Let's

File: docs/tutorial.rst [hunk 6]
```diff
@@ -0,0 +1,987 @@
...
+between her devices.
+
+
+In order to separate data between each user, we will use the default
+user bucket to create our collection:
```
Comment (leplatrem) [inline docs/tutorial.rst]: *personal bucket*

File: docs/tutorial.rst [hunk 7]
```diff
@@ -0,0 +1,987 @@
...
+
+.. code-block:: http
```
Comment (leplatrem) [inline docs/tutorial.rst]: Looks like @ametaireau introduced some sphinx tooling for this

File: docs/tutorial.rst [hunk 8]
```diff
@@ -0,0 +1,987 @@
...
+            ]
+        }
+    }
+
+As soon as the collection has been created, you may want to start to add some tasks in it.
+
+Let start with a really simple data model:
+
+  - ``description``: A string describing the task
+  - ``status``: The status of the task, either ``todo``, ``doing`` or ``done``.
```
Comment (leplatrem) [inline docs/tutorial.rst]: (e.g. ``todo``, ...)

File: docs/tutorial.rst [hunk 9]
```diff
@@ -0,0 +1,987 @@
...
+There you should ask yourself, what happens if another device already
+updated the record in between, will I override its change?
+
+You've got two conflicts resolution behaviors:
+
+- Server wins, in that case the server will reject changes in case
+  something changed on server side.
+- Client wins, in that case the change will override previous changes
+
+The previous call is the Client wins behavior.
```
Comment (leplatrem) [inline docs/tutorial.rst]: previous call above

File: docs/tutorial.rst [hunk 10]
```diff
@@ -0,0 +1,987 @@
...
+
+In case you want the server to prevent you from overridding changes,
```
Comment (leplatrem) [inline docs/tutorial.rst]: In case you want to use the *Server wins* behaviour, you must sent the ...

File: docs/tutorial.rst [hunk 11]
```diff
@@ -0,0 +1,987 @@
...
+you must send the ``If-Match`` header:
+
+Let say, we didn't refresh the server since our first POST and we send
+the ETag we had back then ``"1434641515332"``:
```
Comment (leplatrem) [inline docs/tutorial.rst]: we obtained while fetching the collection

File: docs/tutorial.rst [hunk 12]
```diff
@@ -0,0 +1,987 @@
...
+    Server: nginx/1.4.6 (Ubuntu)
+
+    {
+        "code": 412, 
+        "errno": 114, 
+        "error": "Precondition Failed", 
+        "message": "Resource was modified meanwhile"
+    }
+
+The server reject the modification with a 412 error code.
```
Comment (leplatrem) [inline docs/tutorial.rst]: rejects

File: docs/tutorial.rst [hunk 13]
```diff
@@ -0,0 +1,987 @@
...
+
+In order to fix that, we can either ask for the record we tried to
+update:
```
Comment (leplatrem) [inline docs/tutorial.rst]: In order to merge both version, we can fetch the last version of this single record

File: docs/tutorial.rst [hunk 14]
```diff
@@ -0,0 +1,987 @@
...
+        }, 
+        "permissions": {
+            "write": [
+                "basicauth:10ea4e5fbf849196a4fe8a9c250b737dd5ef17abbeb8f99692d62828465a9823"
+            ]
+        }
+    }
+
+Or we can ask the list of changes from the last time we've synced our
+local store, filtering on the ``_since`` attribute with the value of
```
Comment (leplatrem) [inline docs/tutorial.rst]: list of changes that occured since we have fetched the collection

File: docs/tutorial.rst [hunk 15]
```diff
@@ -0,0 +1,987 @@
...
+            {
+                "description": "Write a tutorial explaining Kinto", 
+                "id": "23eed462-c063-4ae0-81b0-8bf2210bfe86", 
+                "last_modified": 1434642603605, 
+                "status": "doing"
+            }
+        ]
+    }
+
+Now that we've got the list of the record that changed, we can handle the conflict.
```
Comment (leplatrem) [inline docs/tutorial.rst]: records

File: docs/tutorial.rst [hunk 16]
```diff
@@ -0,0 +1,987 @@
...
+We can either do three-way merge (if our changes and server changes on
+the object did not happened on the same fields) or if both objects are
+actually equals.
+
+Or if changes did happened on the same field, we must decide or ask
+the user to decide, which version we have to keep (server version or
+client version).
+
+Then we can try to send back again our modifications using the new
+record ``last_modified`` value:
```
Comment (leplatrem) [inline docs/tutorial.rst]: I would say ETag here

File: docs/tutorial.rst [hunk 17]
```diff
@@ -0,0 +1,987 @@
...
+
+    {
+        "data": {
+            "deleted": true, 
+            "id": "23eed462-c063-4ae0-81b0-8bf2210bfe86", 
+            "last_modified": 1434644823180
+        }
+    }
+
+If you want to sync your local store with record deletion, you can use
```
Comment (leplatrem) [inline docs/tutorial.rst]: If you want to obtain the list of records that were deleted

**Comment by leplatrem** [comment]
(*didn't finish reviewing, back on wednesday*)

File: docs/tutorial.rst (hunk 7 above)
Comment (Natim) [inline docs/tutorial.rst]: httpdomain is more to document RESTAPI's than showing example of usage of it (even if there are example in the documentation of it)

I am not sure it is relevant to use it here.

**Comment by leplatrem** [comment]
Ready to be reviewed/merged

**Pull request merged**